feat: follow up reminders

st3iny opened the pull request:
Implements https://github.com/nextcloud/mail/issues/3996

Diff is containing a lot of tests.

## Testing

1. Disable the delay in the frontend to show follow up reminders right away.
```
Comment or remove the line: src/components/MailboxThread.vue:207
//notAfter.setDate(notAfter.getDate() - 4)
```

2. Reset the classifier job to make it run on the next cron job (and trigger it).
```sql
update oc_jobs set last_run = 0, last_checked = 0 where class like '%FollowUpClassifierJob%';
```


## Architecture

Admins can disable LLM related features completely in the admin/groupware settings section. User can opt-out of this feature through the mail settings modal (bottom left one). It is enabled by default.

The feature will only be applied to new sent emails and not to existing ones. Adding a new account or resetting the cache of the sent mailbox will only trigger the feature for emails not older than 14 days to prevent a job spam.

### Classifier Listener

1. Sending a mail adds it to the sent mailbox.
2. The next sync will pick up changes in the sent mailbox.
3. A classifier job will be queued for each new sent mail after a couple of days.

### Classifier Job

1. Check if a sent message has been responded to by checking its thread for newer messages (meaning there was a reply).
2. If not 1.) Run a LLM free prompt task to check if an email requires a follow-up.
3. If 2) Tag the message with `$follow_up` to notify the frontend. 
3. Otherwise) Do nothing.

### Frontend

- Tagged messages will be shown in the priority mailbox after 4 days.
- Reminders can be followed-up with a dedicated button.
- Reminders can be dismissed through a dedicated button (just deletes the tag).

## Screenshots

| Priority inbox | Thread view | 
| --- | --- |
| ![Screenshot_20240510_130745](https://github.com/nextcloud/mail/assets/1479486/a8625e88-4cf8-4384-9feb-62675a89ec31) | ![image](https://github.com/nextcloud/mail/assets/1479486/28db2377-b3f5-415b-aa82-4e921b2389d7) |

ChristophWurst commented on lib/Listener/FollowUpClassifierListener.php:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	public function handle(Event $event): void {
+		if (!($event instanceof NewMessagesSynchronized)) {
```
Messages are classified when they are first seen. Assuming reliable background sync you will often first see the message added to the cache, then a reply happens, if any.
If we delay this check by 5-6 days we can skip classification of emails that have received a reply by then.

st3iny replied:
That is a good point. It would certainly save a lot of processing power.

ChristophWurst commented:
![Bildschirmfoto vom 2024-05-28 14-53-02](https://github.com/nextcloud/mail/assets/1374172/5046cc7d-65ba-406a-8090-a9f973629807)

classification works. but how am I able to see the nudge? I have modified oc_mail_messages.sent_at to fake time. Do I need anything else?

st3iny commented on lib/BackgroundJob/FollowUpClassifierJob.php:
```diff
@@ -0,0 +1,123 @@
+		}
+
+		$this->logger->debug('Message requires follow-up: ' . $message->getId());
+		$tag = $this->mailManager->createTag('Follow up', '#d77000', $userId);
```
Not necessarily. The `createTag()` method generates a label's value from its display text. Currently, there is no way to specify display name and label value at the same time using this method. I could adjust the API though.

The label is localized in the frontend (if the user didn't rename it manually).

st3iny commented:
> classification works. but how am I able to see the nudge? I have modified oc_mail_messages.sent_at to fake time. Do I need anything else?

The reminder is probably hidden in the frontend and will be shown after 4 days. I added some instructions to the PR's description.

ChristophWurst reviewed (approved):
Tested & works

Design-wise we will need to tone down on the primary elements

![Bildschirmfoto vom 2024-06-06 08-13-02](https://github.com/nextcloud/mail/assets/1374172/ea3d278c-dab4-4eec-aa63-9013f80cda49)

*Disable reminder* could be secondary. 

The test email went to myself. Why do I see follow-up reminders for the outgoing and ingoing message? bug?

st3iny commented:
I lowered the severity of the disable reminder button and moved the follow up button to the bottom. It will now properly replace the reply button. The screenshots in the description are up to date.
